prescient/plugins/tracking.py

```diff
@@ -0,0 +1,1619 @@
+import pandas as pd
+import pyomo.environ as pyo
+from collections import deque
+import os
+
+def get_data_given(df, bus=None, date=None, hour=None, generator=None, fuel_type=None):
+
+    '''
+    This function gets the data out of a pandas dataframe given one or more
+    options, e.g. time.
+
+    Arguments:
+        df: the dataframe we are interested in
+        bus: the bus ID we want [int]/ the bus name we want [str]
+        date: the date we want [str]
+        hour: the hour we want [int]
+        generator: the generator ID [str]
+        fuel_type: generator fuel, e.g. Coal [str]
+    Returns:
+        df: a dataframe that has the information we specified.
+    '''
+
+    # get data given bus id
+    if bus is not None:
+        # in the original rts-gmlc dataset there is a Bus ID col
+        if 'Bus ID' in df.columns:
+            df = df.loc[(df['Bus ID'] == bus)]
+
+        # in the prescient result data we have to extract bus id from other col
+        # e.g. gennerator name col
+        elif 'Generator' in df.columns:
+            # convert the type to str
+            bus = str(bus)
+            # find the rows that starts with the bus name
+            searchrows = df['Generator'].str.startswith(bus)
+            df = df.loc[searchrows,:]
+
+        elif 'Bus' in df.columns:
+            df = df.loc[(df['Bus'] == bus)]
+
+    # get data given date
+    if date is not None:
+        df = df.loc[(df['Date'] == date)]
+
+    # get data given hour
+    if hour is not None:
+        df = df.loc[(df['Hour'] == hour)]
+
+    # get data given hour
+    if generator is not None:
+
+        # Similarly this is for Prescient result data
+        if 'Generator' in df.columns:
+            df = df.loc[(df['Generator'] == generator)]
+        # this is for rts-gmlc dataset
+        elif 'GEN UID' in df.columns:
+            df = df.loc[(df['GEN UID'] == generator)]
+
+    # get data given fuel
+    if fuel_type is not None:
+        df = df.loc[df['Fuel'] == fuel_type]
+
+    return df
+
+
+class DAM_thermal_tracking:
+
+    def __init__(self,rts_gmlc_data_dir,tracking_horizon = 4, generators = None):
+
+        self.generators = generators
+
+        self.model_data = self.assemble_model_data(generator_names = generators, \
+                                                   rts_gmlc_data_dir = rts_gmlc_data_dir)
+        self.model = self.build_tracking_model(generators = generators, \
+                                               track_horizon = tracking_horizon)
+
+        self.result_list = []
+
+        self.daily_stats = {}
+        self.projection = {}
+
+    @staticmethod
+    def assemble_model_data(generator_names, rts_gmlc_data_dir, **kwargs):
```

Comment:
Is this function used in multiple steps (bidding, tracking, etc.)? If so, this is a great reason to make `Model_Data` a standalone class.